authority/provisioners.go

```diff
@@ -19,25 +29,573 @@ func (a *Authority) GetEncryptedKey(kid string) (string, error) {
+	}
+	return nil
+}
+
+// UpdateProvisioner stores an provisioner.Interface to the authority.
+func (a *Authority) UpdateProvisioner(ctx context.Context, nu *linkedca.Provisioner) error {
+	a.adminMutex.Lock()
+	defer a.adminMutex.Unlock()
+
+	certProv, err := ProvisionerToCertificates(nu)
+	if err != nil {
+		return admin.WrapErrorISE(err,
+			"error converting to certificates provisioner from linkedca provisioner")
+	}
+
+	provisionerConfig, err := a.generateProvisionerConfig(ctx)
+	if err != nil {
+		return admin.WrapErrorISE(err, "error generating provisioner config")
+	}
+
+	if err := certProv.Init(*provisionerConfig); err != nil {
+		return admin.WrapErrorISE(err, "error initializing provisioner %s", nu.Name)
+	}
+
+	if err := a.provisioners.Update(certProv); err != nil {
+		return admin.WrapErrorISE(err, "error updating provisioner '%s' in authority cache", nu.Name)
+	}
+	if err := a.adminDB.UpdateProvisioner(ctx, nu); err != nil {
+		if err := a.reloadAdminResources(ctx); err != nil {
+			return admin.WrapErrorISE(err, "error reloading admin resources on failed provisioner update")
+		}
+		return admin.WrapErrorISE(err, "error updating provisioner '%s'", nu.Name)
+	}
+	return nil
+}
+
+// RemoveProvisioner removes an provisioner.Interface from the authority.
+func (a *Authority) RemoveProvisioner(ctx context.Context, id string) error {
+	a.adminMutex.Lock()
+	defer a.adminMutex.Unlock()
+
+	p, ok := a.provisioners.Load(id)
+	if !ok {
+		return admin.NewError(admin.ErrorBadRequestType,
+			"provisioner %s not found", id)
+	}
+
+	provName, provID := p.GetName(), p.GetID()
+	// Validate
+	//  - Check that there will be SUPER_ADMINs that remain after we
+	//    remove this provisioner.
+	if a.admins.SuperCount() == a.admins.SuperCountByProvisioner(provName) {
+		return admin.NewError(admin.ErrorBadRequestType,
+			"cannot remove provisioner %s because no super admins will remain", provName)
+	}
+
+	// Delete all admins associated with the provisioner.
+	admins, ok := a.admins.LoadByProvisioner(provName)
+	if ok {
+		for _, adm := range admins {
+			if err := a.removeAdmin(ctx, adm.Id); err != nil {
+				return admin.WrapErrorISE(err, "error deleting admin %s, as part of provisioner %s deletion", adm.Subject, provName)
+			}
+		}
+	}
+
+	// Remove provisioner from authority caches.
+	if err := a.provisioners.Remove(provID); err != nil {
+		return admin.WrapErrorISE(err, "error removing admin from authority cache")
+	}
+	// Remove provisioner from database.
+	if err := a.adminDB.DeleteProvisioner(ctx, provID); err != nil {
+		if err := a.reloadAdminResources(ctx); err != nil {
+			return admin.WrapErrorISE(err, "error reloading admin resources on failed provisioner remove")
+		}
+		return admin.WrapErrorISE(err, "error deleting provisioner %s", provName)
+	}
+	return nil
+}
+
+func CreateFirstProvisioner(ctx context.Context, db admin.DB, password string) (*linkedca.Provisioner, error) {
```

Comment:
Same comment as above. We're not going to encourage users to turn on `enableAdmin` in the configuration until we button up onboarding. It will just be a hidden flag for now.